### PR TITLE
Add two-class logistic regression

### PR DESCRIPTION
**Pull request by paulkoerbitz**
A simple two-class logistic regression classifier, using argmin's
LBFGS for minimization. The implementation was largely inspired
by Python's sklearn logistic regression.

There are a few points for discussion:
- Should we be maximally generic in the Array and float types? (I think this would be preferable, but it is also more complicated to understand)
- Do we want to pull in a dependency on Argmin? (I think we'll need optimisation throughout linfa and Argmin is great, but its a heavy dependency for logistic regression alone)

### CODE REVIEW

**Comment by bytesnake** [comment]
I'm fine with the additional dependency to `argmin` as long as you don't have to pull it in in order to use other algorithms. Perhaps it's best to have a separate sub-crate for every classifier/regressor and then group them with cargo features. 

Generic over float types is a must (and it's still easier to understand than ndarray types IMO). But sometimes it's easier to define a separate `trait Float` for a whole crate with the necessary definitions, like [here](https://github.com/rust-ml/linfa/blob/master/linfa-reduction/src/lib.rs). 
Cheers and thanks for contributing!

**Comment by paulkoerbitz** [comment]
> Perhaps it's best to have a separate sub-crate for every classifier/regressor and then group them with cargo features.

That's an idea - so I would split this into its own sub-crate 'logistic'? One problem I see here is that one of the next things I wanted to do was Lasso and Ridge (linear) regressions, which also require solvers (so presumably argmin) - should these then go into a different sub-crate as well? It would make the library quite tattered (though I suppose we could re-export things in more coherent batches conditional on features being turned on).

> Generic over float types is a must

Yes, fully agree with this and this is also easy. I also don't mind making things generic in the Endarray type, though this is a bit more work. My main concern though is that it also is harder for the user. OTOH this gives full control and can make things more efficient (why should users be forced to make a copy of an array just because it has the wrong 'type').

Maybe accepting `ArrayView`s where possible is a good approach as I think (?) these can be created from all ndarrays.

**Comment by bytesnake** [comment]
> That's an idea - so I would split this into its own sub-crate 'logistic'? One problem I see here is that one of the next things I wanted to do was Lasso and Ridge (linear) regressions, which also require solvers (so presumably argmin) - should these then go into a different sub-crate as well? It would make the library quite tattered (though I suppose we could re-export things in more coherent batches conditional on features being turned on).
> 
I would just split this into `linfa-logistic` and `linfa-elasticnet`. For LASSO you could use `argmin` or alternatively write your own optimization routine (I think this is what scipy and [SpaSM](http://www.imm.dtu.dk/projects/spasm/)) are doing. 

> > Generic over float types is a must
> 
> Yes, fully agree with this and this is also easy. I also don't mind making things generic in the Endarray type, though this is a bit more work. My main concern though is that it also is harder for the user. OTOH this gives full control and can make things more efficient (why should users be forced to make a copy of an array just because it has the wrong 'type').
> 
> Maybe accepting `ArrayView`s where possible is a good approach as I think (?) these can be created from all ndarrays.

Didn't know OTOH :D if you're using `ArrayBase<impl Data<Elem = A>, Ix?>` you're already accepting `ArrayView`

**Comment by bytesnake** [comment]
actually the presentation for SpaSM has a number of good recipes [here](http://www.imm.dtu.dk/projects/spasm/references/spasm.pdf)

**Comment by paulkoerbitz** [comment]
> actually the presentation for SpaSM has a number of good recipes here
Thanks - looks really interesting, I'll take a deeper look as I'll head into the other regression models (lasso et al.)

This PR is now ready for review - happy if you could take a look.

This time it was actually more challenging to generalise over floats than I thought - Argmin has a lot of bounds on its generic parameters (ArgminOp::Param and ArgminOp::Float) and to implement them I had to wrap Array1 and Float in newtypes and then re-implement all the required traits, which unfortunately makes for a lot of boilerplate code (I actually went ahead and learned a bit about macros to shorten it a bit). That stuff is in argmin_helper.rs. I'd be interested in your thoughts on this and if there is a better approach. I believe if Argmin implemented its traits automatically for all types that fulfill its trait bounds it would just work out of the box - I'll probably file an issue with them and see what they say. For now this was the best I could do.

Another point: I think there is currently no formatting, type-checking and test automation for this repo - I think this would be quite important to enable again to not inadvertently break anything. Happy to help setting this up if you want some help here.

**Comment by bytesnake** [comment]
Nice work, I managed to remove the `FloatWrapper` struct with some bending, but not sure about `ArgminParam`. https://github.com/bytesnake/linfa/commit/a4aa911b4b5c7cbda8e1144aaf364e596b8b867a 
For example the implementation `ArgminMul<F, Array1<F>>` is missing, which I can't implement outside of `argmin` without a wrapper type.

Some tests are still failing though
```
test test::rejects_mismatching_x_y ... ok
test test::rejects_inf_values ... ok
test test::test_logistic_loss ... ok
thread 'test::simple_example_1' panicked at 'assertion failed: `(left == right)`
  left: `[0.68154621024414], shape=[1], strides=[1], layout=CF (0x3), const ndim=1`,
 right: `[0.6817207491013167], shape=[1], strides=[1], layout=CF (0x3), const ndim=1`', linfa-logistic/src/lib.rstest test::test_logistic_grad ... ok
thread 'test::uses_initial_params' panicked at 'assertion failed: res.intercept().abs_diff_eq(&-4.126156756781147, 1e-4)', linfa-logistic/src/lib.rs:714:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
test test::uses_initial_params ... FAILED
:595:9
test test::simple_example_1 ... FAILED
test test::validates_initial_params ... ok
thread 'test::simple_example_2' panicked at 'assertion failed: `(left == right)`
  left: `-4.124856175458871`,
 right: `-4.126156756781147`', linfa-logistic/src/lib.rs:616:9
test test::simple_example_2 ... FAILED
thread 'test::works_with_f32' panicked at 'assertion failed: `(left == right)`
  left: `[0.6815462], shape=[1], strides=[1], layout=CF (0x3), const ndim=1`,
 right: `[0.68260515], shape=[1], strides=[1], layout=CF (0x3), const ndim=1`', linfa-logistic/src/lib.rs:726:9
test test::works_with_f32 ... FAILED

failures:

failures:
    test::simple_example_1
    test::simple_example_2
    test::uses_initial_params
    test::works_with_f32
```

**Comment by paulkoerbitz** [comment]
> Nice work, I managed to remove the FloatWrapper

Oh, that's great, thank you! Slightly embarrassing that I didn't think of this ...

> For example the implementation ArgminMul<F, Array1<F>> 

Yes, this was the motivation for introducing this. I don't see a way around this unfortunately.

> Some tests are still failing though

I think I'm requiring too much precision here, will fix.

**Comment by paulkoerbitz** [comment]
Updated - thanks again for the help with FloatWrapper, much better now!

File: linfa-logistic/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+            params,
+            labels,
+        })
+    }
+}
+
+/// Identify the distinct values of the target 1-d array `y` and associate
+/// the target labels `-1.0` and `1.0` to it.
+fn label_classes<F: Float, A: Data<Elem = F>>(
+    y: &ArrayBase<A, Ix1>,
```
[the remark this answers is not in the document]
Comment (paulkoerbitz) [reply linfa-logistic/src/lib.rs:253]: Yes, that's definitely a good extension. In that case, I think I would also not require this to be an array or even a Vec, but a bound in `IntoIterator` instead.

**Comment by paulkoerbitz** [comment]
Those are very good points - thanks for the review! 

One question I have regarding returning probabilities is how this would work for multi class predictions. Worrying about this might seem premature but its an extension that I was thinking about implementing. I'll take a look at how sklearn handles this, maybe that can be an inspiration. (One option is of course to use a different type for this altogether, maybe this is actually the cleanest option as it mixes concerns the least and then naturally separates these concerns.)

File: linfa-logistic/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulkoerbitz) [reply linfa-logistic/src/lib.rs:253]: Btw, I'm using the terminology the other way around: the user provides _classes_ and I assign them _labels_ (-1.0 or 1.0). 

I see you understood it the other way around. Maybe that's a sign that the terminology isn't that intuitive. OTOH _Two-Class Logistic Regression_ seems more natural than _Two-Label Logistic Regression_.

**Comment by paulkoerbitz** [comment]
OK, fit now takes a non-consuming `IntoIterator` with a `PartialOrd + Clone` bound on the items, which I think is quite flexible. 

Renamed the `predict` method to `predict_classes` and added `predict_probabilties`. 

I'll take another look to update comments, clean up code a bit and maybe add a test, but other than that I think this is good.

**Comment by bytesnake** [comment]
looking good! I was thinking about adding a `Dataset` trait in linfa (similar to Tensorflow), but this would only be a wasted wrapper over `Iterator`. I think you solution is more flexible :)

**Comment by paulkoerbitz** [comment]
Thanks for merging! I was actually planning on updating the comments (pushed this commit when the types were right after some fighting with the type checker - still a bit of a rust noob...). I think I should have marked the PR WIP again - sorry for that.

I'll take another look and file a small follow-up PR if the comments / tests need some updates.

> I was thinking about adding a Dataset trait in linfa (similar to Tensorflow), but this would only be a wasted wrapper over Iterator

I think things like this will make sense as well as taking a step back and making interfaces consistent, maybe introducing some traits (regression & classification come to mind). I imagine this will be a larger discussion and having some concrete examples makes sense - so IMO its still slightly early for this.

**Comment by bytesnake** [comment]
> Thanks for merging! I was actually planning on updating the comments (pushed this commit when the types were right after some fighting with the type checker - still a bit of a rust noob...). I think I should have marked the PR WIP again - sorry for that.
> 
> I'll take another look and file a small follow-up PR if the comments / tests need some updates.

Ah didn't read the former comment properly, sorry :sweat_smile: 
> > I was thinking about adding a Dataset trait in linfa (similar to Tensorflow), but this would only be a wasted wrapper over Iterator


> I think things like this will make sense as well as taking a step back and making interfaces consistent, maybe introducing some traits (regression & classification come to mind). I imagine this will be a larger discussion and having some concrete examples makes sense - so IMO its still slightly early for this.

Yes currently we are collecting algorithms and once we have a good basis, we need to figure out how `partial_fit` etc. traits are supposed to work. Also whether we want to introduce a `Dataset` with additional ability to split itself into batches and how this works with `Iterator`.
